newrelic/node-newrelic
# feat: Added Azure site extension installation scripts

## Description

This adds installation scripts to install the node agent as a site extension in Azure. This PR pairs with #2488 which publishes the node agent as an Azure Site Extension (windows only). 

- install.cmd is the entry file that calls install.ps1
- install.ps1 checks if the user has the node agent installed. If the node agent is already installed, it compares the version of the installed node agent to the latest version. If the versions are different, we install the latest node version. If they're the same versions, we skip installation. If the node agent is not installed, we go straight to installing the latest version (we skipping the check-version function). The installation of the node agent is done by moving the node_module from the extensions bundle (which contains the node agent and dependencies) to the user's node_modules. If that fails for some reason, we install the agent using `npm`. 
- uninstall.cmd uninstalls the agent using `npm uninstall` and it also checks for any empty `@` directories in `node_modues` and deletes them if they are empty. This is needed because `npm uninstall` leaves empty `@` directories and when a user reinstalls the agent site extension, our installation script fails because it can't move packages to empty `@` directories. 

Note: [this PR](#2488) will contain the readme for this extension. 

## How to Test

Create a sample web app in Azure and install the Node Agent site extension. Make sure to add your license key to your environment variables. 

## Screenshots of the test extension installation working
### Shows the extension directory added after installation with all the necessary files:
<img width="1121" alt="testExtension directory in Azure" src="https://github.com/user-attachments/assets/8edaa89f-0945-4e41-bca3-6d6d5d7e7152">

### wwwroot showing the agent log and node agent version:
<img width="1237" alt="wwwroot showing files" src="https://github.com/user-attachments/assets/41338ef8-9054-47da-bf8f-0054665506e1">

### sample test app in NR1 showing data flowing and instrumentation working:
<img width="1505" alt="data flowing in NR1" src="https://github.com/user-attachments/assets/16dcdf34-a916-4921-8ca0-5ea901e904b3">

#### Node VMs dashboard 
<img width="1441" alt="Screenshot 2024-09-03 at 3 32 05 PM" src="https://github.com/user-attachments/assets/ee4c09b1-a872-40ae-8ce9-22dc41dd14d9">

#### `code.` attributes in spans
<img width="1449" alt="Screenshot 2024-09-03 at 3 31 52 PM" src="https://github.com/user-attachments/assets/4c8798be-5174-4fe9-b3aa-7f4d61831f31">

## Related Issues

Closes #2442 
Relates to #2488
